core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncRequestBody.java

```diff
@@ -239,12 +253,20 @@ private void readData() {
             inputChannel.read(buffer, position.get(), buffer, new CompletionHandler<Integer, ByteBuffer>() {
                 @Override
                 public void completed(Integer result, ByteBuffer attachment) {
-
                     if (result > 0) {
                         attachment.flip();
-                        position.addAndGet(attachment.remaining());
+
+                        int readBytes = attachment.remaining();
+                        position.addAndGet(readBytes);
+                        remainingBytes.addAndGet(-readBytes);
+
                         signalOnNext(attachment);
 
+                        if (remainingBytes.get() == 0) {
+                            signalOnComplete();
+                            closeFile();
```

Comment:
Would it make sense to call closeFile first in case the the implementation of `subscriber#onComplete` can take a long time to finish or even gets blocked (although not likely)?

Reply:
I think that's fine; fwiw this is the same call order as in `signalOnComplete()`: https://github.com/aws/aws-sdk-java-v2/blob/2a663fc6526ce0004bd023a87d26c7da24eae9f7/core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/FileAsyncRequestBody.java#L258-L259

I'll change both